File: test/unit/python/test_reporter_manager.py
from pxr import Usd
import unf


def test_manager_create():
    """Create a single reporter manager per stage."""
    stage = Usd.Stage.CreateInMemory()
    manager1 = unf.ReporterManager.Create(stage)
    assert unf.ReporterManager.GetNumManagers() == 1

    manager2 = unf.ReporterManager.Create(stage)
    assert unf.ReporterManager.GetNumManagers() == 1
    assert manager1 == manager2

    otherStage = Usd.Stage.CreateInMemory()
    manager3 = unf.ReporterManager.Create(otherStage)
    assert unf.ReporterManager.GetNumManagers() == 2
    assert manager1 != manager3

def test_manager_stage_expired():
    """Clean manager when stage is expired."""
    stage1 = Usd.Stage.CreateInMemory()
    manager1 = unf.ReporterManager.Create(stage1)

    # Stage is destroyed, but manager reference is kept in registry
    del stage1
    assert manager1.expired is False
    assert unf.ReporterManager.GetNumManagers() == 1

    # Registry reference is removed when a manager is added
    stage2 = Usd.Stage.CreateInMemory()
    manager2 = unf.ReporterManager.Create(stage2)
    assert manager1.expired is True
    assert manager2.expired is False
    assert unf.ReporterManager.GetNumManagers() == 1

def test_manager_reset():
    """Explicitly reset manager."""
    stage1 = Usd.Stage.CreateInMemory()
    manager1 = unf.ReporterManager.Create(stage1)

    stage2 = Usd.Stage.CreateInMemory()
    manager2 = unf.ReporterManager.Create(stage2)

    # Explicitly reset one manager
    manager1.Reset()
    assert manager1.expired is True
    assert manager2.expired is False

def test_manager_reset_all():
    """Explicitly reset all managers."""
    stage1 = Usd.Stage.CreateInMemory()
    manager1 = unf.ReporterManager.Create(stage1)

    stage2 = Usd.Stage.CreateInMemory()
    manager2 = unf.ReporterManager.Create(stage2)

    # Explicitly reset one manager
    unf.ReporterManager.ResetAll()
    assert manager1.expired is True
    assert manager2.expired is True

File: test/unit/testReporterManager.cpp
#include <unf/reporterManager.h>

#include <gtest/gtest.h>
#include <pxr/pxr.h>

TEST(ReporterManagerTest, CreateMultiple)
{
    auto stage = PXR_NS::UsdStage::CreateInMemory();
    auto manager1 = unf::ReporterManager::Create(stage);
    ASSERT_EQ(manager1->GetCurrentCount(), 1);
    ASSERT_EQ(unf::ReporterManager::GetNumManagers(), 1);

    auto manager2 = unf::ReporterManager::Create(stage);
    ASSERT_EQ(manager1->GetCurrentCount(), 1);
    ASSERT_EQ(manager2->GetCurrentCount(), 1);
    ASSERT_EQ(unf::ReporterManager::GetNumManagers(), 1);

    // Managers referencing the same stage are identical
    ASSERT_EQ(manager1, manager2);

    auto otherStage = PXR_NS::UsdStage::CreateInMemory();
    auto manager3 = unf::ReporterManager::Create(otherStage);
    ASSERT_EQ(manager1->GetCurrentCount(), 1);
    ASSERT_EQ(manager2->GetCurrentCount(), 1);
    ASSERT_EQ(manager3->GetCurrentCount(), 1);
    ASSERT_EQ(unf::ReporterManager::GetNumManagers(), 2);

    // Managers referencing distinct stages are different
    ASSERT_NE(manager1, manager3);
}

TEST(ReporterManagerTest, StageExpired1)
{
    auto stage1 = PXR_NS::UsdStage::CreateInMemory();
    auto manager1 = unf::ReporterManager::Create(stage1);

    // Stage is destroyed, but manager reference is kept in registry
    stage1.Reset();
    ASSERT_FALSE(manager1.IsExpired());
    ASSERT_EQ(unf::ReporterManager::GetNumManagers(), 1);

    // Registry reference is removed when a manager is added
    auto stage2 = PXR_NS::UsdStage::CreateInMemory();
    auto manager2 = unf::ReporterManager::Create(stage2);
    ASSERT_TRUE(manager1.IsExpired());
    ASSERT_FALSE(manager2.IsExpired());
    ASSERT_EQ(unf::ReporterManager::GetNumManagers(), 1);
}

TEST(ReporterManagerTest, StageExpired2)
{
    unf::ReporterManagerWeakPtr manager1;

    {
        auto stage1 = PXR_NS::UsdStage::CreateInMemory();
        manager1 = unf::ReporterManager::Create(stage1);
    }

    // Stage is destroyed, but manager reference is kept in registry
    ASSERT_FALSE(manager1.IsExpired());
    ASSERT_EQ(unf::ReporterManager::GetNumManagers(), 1);

    // Registry reference is removed when a manager is added
    auto stage2 = PXR_NS::UsdStage::CreateInMemory();
    auto manager2 = unf::ReporterManager::Create(stage2);
    ASSERT_TRUE(manager1.IsExpired());
    ASSERT_FALSE(manager2.IsExpired());
    ASSERT_EQ(unf::ReporterManager::GetNumManagers(), 1);
}

TEST(ReporterManagerTest, Reset1)
{
    auto stage1 = PXR_NS::UsdStage::CreateInMemory();
    auto manager1 = unf::ReporterManager::Create(stage1);

    auto stage2 = PXR_NS::UsdStage::CreateInMemory();
    auto manager2 = unf::ReporterManager::Create(stage2);

    // Explicitly reset one manager
    manager1->Reset();
    ASSERT_TRUE(manager1.IsExpired());
    ASSERT_FALSE(manager2.IsExpired());
}

TEST(ReporterManagerTest, Reset2)
{
    auto stage1 = PXR_NS::UsdStage::CreateInMemory();
    auto manager1 = unf::ReporterManager::Create(stage1);

    auto stage2 = PXR_NS::UsdStage::CreateInMemory();
    auto manager2 = unf::ReporterManager::Create(stage2);

    // Explicitly reset one manager from associated stage
    unf::ReporterManager::Reset(stage1);
    ASSERT_TRUE(manager1.IsExpired());
    ASSERT_FALSE(manager2.IsExpired());
}

TEST(ReporterManagerTest, ResetAll)
{
    auto stage1 = PXR_NS::UsdStage::CreateInMemory();
    auto manager1 = unf::ReporterManager::Create(stage1);

    auto stage2 = PXR_NS::UsdStage::CreateInMemory();
    auto manager2 = unf::ReporterManager::Create(stage2);

    // Explicitly reset all managers
    unf::ReporterManager::ResetAll();
    ASSERT_TRUE(manager1.IsExpired());
    ASSERT_TRUE(manager2.IsExpired());
}
